src/java_bytecode/java_string_library_preprocess.cpp

```diff
@@ -1169,6 +1174,302 @@ codet java_string_library_preprocesst::make_string_to_char_array_code(
+  for(const auto & comp : structured_type.components())
+  {
+    const irep_idt &name=comp.get_name();
+    exprt field_expr=field_exprs.front();
+    field_exprs.pop_front();
+
+    if(name=="string_expr")
+    {
+      pointer_typet string_pointer(symbol_typet("java::java.lang.String"));
+      typecast_exprt arg_i_as_string(arg_i, string_pointer);
+      code_not_null.add(code_assign_java_string_to_string_expr(
+        to_string_expr(field_expr), arg_i_as_string , symbol_table));
+      exprt arg_i_string_expr_sym=fresh_string_expr_symbol(
+        loc, symbol_table, code_not_null);
+      code_not_null.add(code_assignt(
+        arg_i_string_expr_sym, to_string_expr(field_expr)));
+    }
+    else if(name==ID_int || name==ID_float || name==ID_char || name==ID_boolean)
+    {
+      exprt value=get_primitive_value_of_object(
+        arg_i, name, loc, symbol_table, code_not_null);
+      code_not_null.add(code_assignt(field_expr, value));
+    }
+    else
+    {
+      // TODO: date_time and hash_code not implemented
+    }
+  }
+
+  code.add(code_not_null);
+  return arg_i_struct;
+}
+
+/// Used to provide code for the Java String.format function.
+///
+/// TODO: date_time and hash code are not implemented, and we set a limit of
+/// 10 arguments
+/// \param type: type of the function call
+/// \param loc: location in the program_invocation_name
+/// \param symbol_table: symbol table
+/// \return Code implementing the Java String.format function.
+/// Since the exact class of the arguments is not known, we give as argument
+/// to the internal format function a structure containing the different
+/// possible types.
+codet java_string_library_preprocesst::make_string_format_code(
+  const code_typet &type,
+  const source_locationt &loc,
+  symbol_tablet &symbol_table)
+{
+  code_blockt code;
+  exprt::operandst args=process_parameters(
+    type.parameters(), loc, symbol_table, code);
+
+  // We will process arguments so that if it has to be casted into a string
+  // then a string_exprt will be available in the arguments
+  std::vector<exprt> processed_args;
+
+  // The argument can be:
+  // a string, a integer, a floating point, a character, a boolean,
+  // an object of which we take the hash code, or a date/time
+  struct_typet structured_type;
+  structured_type.components().emplace_back("string_expr", refined_string_type);
+  structured_type.components().emplace_back(ID_int, java_int_type());
+  structured_type.components().emplace_back(ID_float, java_float_type());
+  structured_type.components().emplace_back(ID_char, java_char_type());
+  structured_type.components().emplace_back(ID_boolean, java_boolean_type());
+  // TODO: hash_code not implemented for now
+  structured_type.components().emplace_back("hashcode", java_int_type());
+  // TODO: date_time type not implemented for now
+  structured_type.components().emplace_back("date_time", java_int_type());
+
+  // Arbitrary limit of 10 arguments for format
```

Comment:
I wonder if these magical numbers should all be put in the same place (e.g. `#define MAX_FORMAT_ARGS 10` at the top of the file). Otherwise it could become difficult to remember all the specific values that we scattered in the code.